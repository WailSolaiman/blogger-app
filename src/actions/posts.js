import database from '../firebase/firebase';

//create post
export const createPost = (post) => ({
    type: 'CREATE_POST',
    post
});

export const startCreatePost = (postData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const { title = '', content= '', createdAt = 0 } = postData;
        const post = { title, content, createdAt };
        return database.ref(`users/${uid}/posts`).push(post).then((ref) => {
            dispatch(createPost({
              pid: ref.key,
              ...post
            }));
        });
    };
};
//edit post
export const editPost = (pid, updates) => ({
    type: 'EDIT_POST',
    pid,
    updates
});

export const startEditPost = (pid, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/posts/${pid}`).update(updates).then(() => {
            dispatch(editPost(pid, updates));
        });
    };
};
//remove post
export const removePost = (pid) => ({
    type: 'REMOVE_POST',
    pid
});

export const startRemovePost = (pid) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/posts/${pid}`).remove().then(() => {
            dispatch(removePost(pid));
        });
    };
};
// set posts
export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = (posts) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
            const posts = [];
            snapshot.forEach((childSnapshot) => {
              posts.push({
                pid: childSnapshot.key,
                ...childSnapshot.val()
              });
            });
            dispatch(setPosts(posts));
        });
    };
};