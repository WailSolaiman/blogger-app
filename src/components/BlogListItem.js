import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const BlogListItem = ({pid, title, content, createdAt}) => {
    return (
        <Link className="list-item" to={`/edit/${pid}`}>
            <h1>{title}</h1>
            <p><small><strong>{moment(createdAt).format('MMMM Do, YYYY')}</strong></small></p>
            <p>{content}</p>
        </Link>
    );
}

export default BlogListItem;