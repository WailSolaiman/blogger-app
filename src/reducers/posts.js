export default (state = [], action) => {
    switch(action.type) {
        case 'CREATE_POST':
            return [
                ...state,
                action.post
            ];
        case 'EDIT_POST':
            return state.map((post) => {
                if (post.pid === action.pid) {
                    return {
                        ...post,
                        ...action.updates
                    };
                }
                else {
                    return post;
                }
            });
        case 'REMOVE_POST':
            return state.filter((post) => post.pid !== action.pid);
        case 'SET_POSTS':
            return action.posts;
        default:
            return state;
    }
};