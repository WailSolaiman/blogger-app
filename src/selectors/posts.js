import moment from 'moment';

export default (posts, {text, sortBy}) => {
    return posts.filter((post) => (
        post.title.toLowerCase().includes(text.toLowerCase())
    ))
    .sort((a, b) => {
        if (sortBy === 'date') {
          return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'title') {
          return a.title > b.title ? 1 : -1;
        }
    });
};