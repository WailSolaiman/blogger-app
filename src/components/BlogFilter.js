import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setTextFilter, sortByTitle, sortByDate } from '../actions/filters';

const BlogFilter = (props) => {
    return (
        <div className="content-container">
            <div className="input-group">
                <div className="input-group__item">
                    <input 
                        type="text"
                        placeholder="Search in Blogger"
                        value={props.filters.text}
                        onChange={(event) => {
                            props.setTextFilter(event.target.value)
                        }}
                    />
                </div>
                <div className="input-group__item">
                    <select 
                        value={props.filters.sortBy}
                        onChange={(event) => {
                            if (event.target.value === 'title') {
                                props.sortByTitle();
                            } else if (event.target.value === 'date') {
                                props.sortByDate();
                            }
                        }}
                    >
                        <option value="title">Sort posts by title</option>
                        <option value="date">Sort by by date</option>
                    </select>
                </div>
                <div className="input-group__item">
                    <Link className="button button--link button--text-black" to="/create"><p>Add Post</p></Link>
                </div>
            </div>
        </div>
        
    );
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByTitle: () => dispatch(sortByTitle()),
    sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogFilter);