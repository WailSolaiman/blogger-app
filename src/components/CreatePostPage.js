import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startCreatePost } from '../actions/posts';

export const CreatePostPage = (props) => {
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add new Post</h1>
                </div>
            </div>
            <div className="content-container">
                <PostForm onSubmit={(post) => {
                    props.startCreatePost(post);
                    props.history.push('/dashboard');
                }} />
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startCreatePost: (post) => dispatch(startCreatePost(post))
});

export default connect(undefined, mapDispatchToProps)(CreatePostPage);