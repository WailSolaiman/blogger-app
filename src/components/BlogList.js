import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';
import selectPosts from '../selectors/posts';

const BlogList = (props) => (
    <div className="content-container">
        {
            props.posts.map((post) => {
                return <BlogListItem key={post.pid} {...post} />
            })
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        posts: selectPosts(state.posts, state.filters)
    };
};

export default connect(mapStateToProps)(BlogList);