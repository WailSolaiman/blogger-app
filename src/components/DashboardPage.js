import React from 'react';
import BlogFilter from './BlogFilter';
import BlogList from './BlogList';

const DashboardPage = () => (
  <div>
    <BlogFilter />
    <BlogList />
  </div>
);

export default DashboardPage;
