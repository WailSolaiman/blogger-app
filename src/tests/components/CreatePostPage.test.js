import React from 'react';
import { shallow } from 'enzyme';
import { CreatePostPage }  from '../../components/CreatePostPage';
import posts from '../fixtures/posts';

let createPost, history, wrapper;

beforeEach(() => {
    createPost = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<CreatePostPage createPost={createPost} history={history} />);
});

test('should render component CreatePostPage succefully', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle submit corectly', () => {
    wrapper.find('PostForm').prop('onSubmit')(posts[1]);
    expect(history.push).toHaveBeenLastCalledWith('/dashboard');
    expect(createPost).toHaveBeenLastCalledWith(posts[1]);
});