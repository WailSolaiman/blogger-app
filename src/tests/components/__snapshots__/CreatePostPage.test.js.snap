// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component CreatePostPage succefully 1`] = `
<PostForm
  onSubmit={[Function]}
/>
`;
